pulp_ansible/app/tasks/signature.py

```diff
@@ -22,6 +23,38 @@
 from pulpcore.plugin.models import SigningService, ProgressReport
 from pulpcore.plugin.sync import sync_to_async_iterable, sync_to_async
 from pulp_ansible.app.tasks.utils import get_file_obj_from_tarball
+from rest_framework import serializers
+
+
+def verify_signature_upload(data):
+    """The task code for verifying collection signature upload."""
+    file = data["file"]
+    sig_data = file.read()
+    file.seek(0)
+    collection = data["signed_collection"]
+    keyring = None
+    if (repository := data.get("repository")) and repository.keyring:
+        keyring = repository.keyring
+    gpg = gnupg.GPG(keyring=keyring)
+    artifact = collection.contentartifact_set.select_related("artifact").first().artifact.file.name
+    artifact_file = storage.open(artifact)
+    with tarfile.open(fileobj=artifact_file, mode="r") as tar:
+        manifest = get_file_obj_from_tarball(tar, "MANIFEST.json", artifact_file)
+        with open("MANIFEST.json", mode="wb") as m:
+            m.write(manifest.read())
+        verified = gpg.verify_file(file, m.name)
+
+    if verified.trust_level is None or verified.trust_level < verified.TRUST_FULLY:
+        # Skip verification if repository isn't specified, or it doesn't have a keyring attached
+        if verified.fingerprint is None or keyring is not None:
+            raise serializers.ValidationError(
+                _("Signature verification failed: {}").format(verified.status)
+            )
+
+    data["data"] = sig_data
+    data["digest"] = file.hashers["sha256"].hexdigest()
+    data["pubkey_fingerprint"] = verified.fingerprint
```

Comment:
Ahh, slick. This is why we can make that field read only...
But it will be None, if we hit the condition in line 49. Would it be better if we could fall back to validation against a global keyring there?
Consider this:
1. you upload the signature; no repo -> kfp=None
2. you add the signature to a repo with a keyring -> kfp=None, because content is immutable
But it should create the same result as
1. you upload the signature to repo with keyring -> kfp is not None

Alternatively only allow upload in cases we can actually verify the sig.

Reply:
The comment I left isn't very indicative of the actual behavior of the check. `verified.fingerprint` will always be set if the signature is a real PGP signature. Even if the signature isn't valid because it's expired or for the wrong file and even if the keyring doesn't contain the public key of the signature the field would be set. `verified.fingerprint` should only be `None` if the signature is beyond messed up and the check on line 49 would ensure that a `ValidationError` is raised. Also, if the keyring isn't set gnupg will use the default keyring from the default homedir of GPG, usually `~/.gnupg/pubring.kbx`. Also, one last thing, `pubkey_fingerprint` can not be set to `None`, it will result in a database error upon creation. So even if it gets passed my checks somehow, the user still wouldn't be able to create a signature that doesn't have a `pubkey_fingerprint`.